smithy-linters/src/main/java/software/amazon/smithy/linters/WordBoundaryMatcher.java

```diff
@@ -0,0 +1,168 @@
+                if (i == 0 || i == pattern.length() - 1 || previousSpace) {
+                    throw new IllegalArgumentException("Invalid spaces in word boundary search: " + pattern);
+                } else {
+                    previousSpace = true;
+                }
+            } else if (!Character.isLetterOrDigit(c)) {
+                throw new IllegalArgumentException(
+                        "Invalid non-alphanumeric characters in word boundary search:" + pattern);
+            } else {
+                previousSpace = false;
+            }
+        }
+
+        return result.toString();
+    }
+
+    // Rather than return a list of words, this method canonicalizes words into a lowercase, space-delimited string.
+    // This allows substring checks to be used to determine if a search is found in the delimited words.
+    // Adapted from Apache Commons Lang3: https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/StringUtils.java#L7457
+    private static String splitWords(String str) {
+        if (str.isEmpty()) {
+            return "";
+        }
+
+        final StringBuilder result = new StringBuilder();
+        int tokenStart = 0;
+        int currentType = Character.getType(str.charAt(tokenStart));
+
+        for (int pos = tokenStart + 1; pos < str.length(); pos++) {
+            final char c = str.charAt(pos);
+            final int type = Character.getType(c);
+            if (type == currentType) {
+                continue;
+            }
+            if (type == Character.LOWERCASE_LETTER && currentType == Character.UPPERCASE_LETTER) {
+                final int newTokenStart = pos - 1;
+                if (newTokenStart != tokenStart) {
+                    addLowerCaseStringToBuilder(result, str, tokenStart, newTokenStart - tokenStart);
+                    result.append(' ');
+                    tokenStart = newTokenStart;
+                }
+            } else {
+                // Skip character groupings that are delimiters. We just want letters and numbers.
+                if (Character.isLetterOrDigit(str.charAt(pos - 1))) {
+                    addLowerCaseStringToBuilder(result, str, tokenStart, pos - tokenStart);
+                    result.append(' ');
+                }
+                tokenStart = pos;
+            }
+            currentType = type;
+        }
+
+        if (Character.isLetterOrDigit(str.charAt(tokenStart))) {
+            // Add the last segment if it's a letter or number.
+            addLowerCaseStringToBuilder(result, str, tokenStart, str.length() - tokenStart);
+        } else {
+            // Since the last segment is ignored, remove the trailing space.
+            result.setLength(result.length() - 1);
+        }
+
+        return result.toString();
+    }
+
+    private static void addLowerCaseStringToBuilder(StringBuilder result, String str, int start, int count) {
```

Comment:
Small nit, I think that we can simplify the callers and a bit the function if we instead take the `endIndex` instead of the `count`.

Reply:
I was mimicking what java.lang.String's constructor does here:

```
public String(char value[], int offset, int count) {
```

It's internal to the class, so probably not worth changing IMO